Sources/PostgresNIO/New/Messages/DataRow.swift

```diff
@@ -1,34 +1,132 @@
+    var columnCount: Int16
+    
+    var bytes: ByteBuffer
+    
+    static func decode(from buffer: inout ByteBuffer) throws -> Self {
+        try buffer.ensureAtLeastNBytesRemaining(2)
+        let columnCount = buffer.readInteger(as: Int16.self)!
+        let firstColumnIndex = buffer.readerIndex
         
-        static func decode(from buffer: inout ByteBuffer) throws -> Self {
+        for _ in 0..<columnCount {
             try buffer.ensureAtLeastNBytesRemaining(2)
-            let columnCount = buffer.readInteger(as: Int16.self)!
+            let bufferLength = Int(buffer.readInteger(as: Int32.self)!)
             
-            var result = [ByteBuffer?]()
-            result.reserveCapacity(Int(columnCount))
-            
-            for _ in 0..<columnCount {
-                try buffer.ensureAtLeastNBytesRemaining(2)
-                let bufferLength = Int(buffer.readInteger(as: Int32.self)!)
-                
-                guard bufferLength >= 0 else {
-                    result.append(nil)
-                    continue
-                }
-                
-                try buffer.ensureAtLeastNBytesRemaining(bufferLength)
-                let columnBuffer = buffer.readSlice(length: Int(bufferLength))!
-                
-                result.append(columnBuffer)
+            guard bufferLength >= 0 else {
+                // if buffer length is negative, this means that the value is null
+                continue
             }
             
-            return DataRow(columns: result)
+            try buffer.ensureAtLeastNBytesRemaining(bufferLength)
+            buffer.moveReaderIndex(forwardBy: bufferLength)
+        }
+        
+        try buffer.ensureExactNBytesRemaining(0)
+        
+        buffer.moveReaderIndex(to: firstColumnIndex)
+        let columnSlice = buffer.readSlice(length: buffer.readableBytes)!
+        return DataRow(columnCount: columnCount, bytes: columnSlice)
+    }
+}
+
+extension DataRow: Sequence {
+    typealias Element = ByteBuffer?
+    
+    // There is no contiguous storage available... Sadly
+    func withContiguousStorageIfAvailable<R>(_ body: (UnsafeBufferPointer<ByteBuffer?>) throws -> R) rethrows -> R? {
+        nil
+    }
+}
+
+extension DataRow: Collection {
+    
+    struct ColumnIndex: Comparable {
+        var offset: Int
+        
+        init(_ index: Int) {
+            self.offset = index
+        }
+        
+        static func == (lhs: Self, rhs: Self) -> Bool {
+            lhs.offset == rhs.offset
+        }
+        
+        static func < (lhs: Self, rhs: Self) -> Bool {
+            lhs.offset < rhs.offset
+        }
+
+        static func <= (lhs: Self, rhs: Self) -> Bool {
+            lhs.offset <= rhs.offset
+        }
+
+        static func >= (lhs: Self, rhs: Self) -> Bool {
+            lhs.offset >= rhs.offset
         }
+
+        static func > (lhs: Self, rhs: Self) -> Bool {
+            lhs.offset > rhs.offset
+        }
```

Comment:
Doesn't the runtime's default implementations of these operators for `Comparable` already provide these in terms of `<` and `==`? Also won't the compiler synthesize the `Equatable` conformance?